tests.md

```diff
@@ -0,0 +1,168 @@
+of a reusable Terraform module always uses some provider, and the provider always speaks to some cloud's REST API.
+Typically the provider+cloud run time is much longer, orders of magnitude longer, than the module run time. Realistic
+example may be 100 seconds versus less than 1 second.
+
+To enable running multiple test cases in much shorter time than that, one can either mock the provider
+or mock the cloud's REST API. Such a "mocked" component basically either returns hardcoded results or an error
+when anything unexpected happens, so it only takes microseconds or so. However such method requires either:
+
+- the Terraform team to be versed in Go language, to be able to implement their own entire mock provider,
+- or, alternatively, the Terraform team to be versed enough in some language to implement their own cloud REST API,
+to have intimate knowledge of the original REST API of that cloud, and to be able to configure the provider to not to
+speak to the real cloud anymore.
+
+Such large investment appears for now to be only possible through an organization-level decision, not a team-level one.
+So the default is not to do the testing this way. Instead we resort to mainly `terraform apply` to the real cloud,
+which one would normally call integration testing. Just as for example Go/Python projects' integration tests which could
+speak to the third-party API or spin up the database instance.
+
+As a corollary to this, the terms "Unit Test" and "Integration Test" should be used with care with Terraform. If you
+mean something else than this section, clarify exactly what are your definitions and how the distinction matters in your
+context.
+
+The only test-like action which is designed to be quite quick with Terraform is the good old `terraform plan`. It's
+worth running, but it can hardly be considered an exhaustive test. The Plan does not dynamically generate the
+identifiers, instead reporting them as `id = (known after apply)`, so object-to-object relations cannot be verified
+at all.
+
+## Test Mainly An Empty Terraform State
+
+The assumption is that we mainly run tests starting from an empty tfstate. For most kinds of tests, maintaining
+any permanent cloud objects and the associated Terraform state would be quite hard. The main reason is not the cost,
+but the fact that breaking changes in the code tend to destroy/recreate a lot of such cloud objects. Such destroys can
+quite often enter a conflict of some kind preventing any further Terraform modification to bring back the well-known
+state for subsequent tests.
+
+## Test Cases: Simple Or Complex?
+
+Per the previous sections, we resort to using real Apply in tests, and it usually brings in some cloud cost. (The Apply
+means the stage after the plan has been already shown to the user, approved, and when it is being deployed to the real
+world.)
+
+If quick and cheap unit tests were possible, it would be preferable to have each test case just cleanly one thing
```

Comment:
```suggestion
If quick and cheap unit tests were possible, it would be preferable to have each test case responsible for one thing
```
Nit: I'm not sure about the intention behind the sentence in the original version - this is how I understand what you meant? Ignore me if I'm wrong of course :)